### ISSUE TITLE
Followup ML

### ISSUE THREAD

**Issue by Carreau** [jdfreder/single-cell-rtcollab#3]
@minrk Said:

> I know Jason has been talking about application state with an eye toward what should be synced. Do you want to make a draft of the 'live notebook' state that we should be synchronizing, and how it differs from the disk format?

Don't really care, for a first step stop checking and assuming server side that the notebook format is respected. 
It should be an implementation detail. 

Then be more strict about the source of cell, it should be either a multiline string, or a list of string, 
but we choose and stick to it. 

Preference for list of string as the is how our curent on-disk format is. 

For the In-Memory format  I concur with (Jon and Kyle)'s  description of a 

ID -> Cell  Mapping, and cellOrder. Note that the CellOrder is likely a list of ID, I guess, which is "just" a simple case of DAG, 
as we discussed in 2013 on the Mailing list. 

My **guess** is that will make life of postgres and some other  content manager much easier, 
it also technically allows you to open just one cell in a normal text editor. 

Anyway, I'll be short on time to contribute to that until end of next week, but if there is a real move forward that would be nice. 

I really think that there is actual work that we can start doing on the notebook server itself that could help in the future.

**Comment by jdfreder** [comment]
Thanks for the post @Carreau !  What are your thoughts about widget state, which lives in the kernel?  Should it be duplicated to the state share server (notebook server)?

**Comment by Carreau** [comment]
> Should it be duplicated to the state share server (notebook server)?

No, but likely widgets ID's, should be stored on the server side. 

But typically things like [that](https://github.com/jupyter/notebook/blob/dcd9027af220ad24c8b5b56d41754bce4b00403c/notebook/static/notebook/js/codecell.js#L347) :+1: 

```
reply : $.proxy(this._handle_execute_reply, this)
```

should be replace by roughly : 

```
reply: {'_handle_execute_reply', this.cell_id}
```

Which **can** be synced on the sever side, meaning that if you refresh the page a pending reply have at least some chance of surviving a page reload (unless the reply arrives while the page is refreshing).

**Comment by rgbkrk** [comment]
> For the In-Memory format I concur with (Jon and Kyle)'s description of a
> ID -> Cell Mapping, and cellOrder. Note that the CellOrder is likely a list of ID

That's correct. The `cellOrder` is a list of `id`s and the `cellMap` is `id -> cell`. `id`s are v4 UUIDs.

A little bit off topic for a single cell exploration, though I'm happy to talk about the notebook model we're experimenting with. It would be good to have others think about and use it in practice.

**Comment by minrk** [comment]
>  for a first step stop checking and assuming server side that the notebook format is respected. 

Are you saying we shouldn't even check that we are writing valid notebooks to disk? That's the only thing we do check now. The in-memory format being discussed that should be different is never validated in the current system.

**Comment by williamstein** [comment]
Of related interest -- may I ask if there is a description somewhere of the
definition of a valid notebook file.   I spent a lot of time thinking about
exactly this yesterday in my Jupyter sync rewrite.   I like to assume the
files can get corrupted, and things go arbitrarily to hell, then make the
application fix things that get corrupted.  My users are frequently
confused students who do all manner of crazy random things, and in real
life there are no guarantees.  Etc.    I also wrote a possibly more careful
version of the notebook fromJSON yesterday, which mutates the notebook
(preserving cursors, etc.), rather than just obliterating and recreating
the whole thing.

On Thu, Mar 10, 2016 at 11:45 PM, Min RK notifications@github.com wrote:

> for a first step stop checking and assuming server side that the notebook
> format is respected.
> 
> Are you saying we shouldn't even check that we are writing valid notebooks
> to disk? That's the only thing we do check now. The in-memory format being
> discussed that should be different is never validated in the current system.

## 

William (http://wstein.org)

**Comment by jasongrout** [comment]
There is a vague description in the docs at https://nbformat.readthedocs.org/en/latest/format_description.html. I also wrote up an unofficial typescript spec for it, if you're interested.

**Comment by jasongrout** [comment]
@williamstein - the notebook disk format doesn't have cursor information, etc. How are you adding information like that to the current notebook (i.e., how is your in-memory model different than the disk model)?

**Comment by rgbkrk** [comment]
I put together an `npm` installable version of the notebook schema here: https://github.com/rgbkrk/nbschema

**Comment by jasongrout** [comment]
Thanks, Kyle!  Right, of course - the actual nbformat repo has the formal spec.

**Comment by jasongrout** [comment]
We should note that, as Matthias hinted at above, the formal spec (at https://github.com/rgbkrk/nbschema/blob/master/v4/nbformat.v4.schema.json#L360) has the multiline string as either a string or a list of strings. The model that the server gives in response to API requests always makes this a string, but it is stored on disk as a list of strings (so that diffs are more understandable, IIRC).

**Comment by williamstein** [comment]
> the notebook disk format doesn't have cursor information, etc. How are you adding information like that to the current notebook
> 
>  (i.e., how is your in-memory model different than the disk model)?

YES, it's completely different.

User cursors are stored in a completely different table, which has nothing to do with either model.  Periodically when a user moves their cursor(s), that table gets updated with the latest location.  That table is synchronized between all clients.  There's no need to have the history of cursor locations (like there is to have the history of the document).   I emphasize _cursors_ since Jupyter (via Codemirror5) fully supports multiple cursors.   In the upcoming SMC release, we'll be showing multiple cursors for other users too.... finally.

**Comment by Carreau** [comment]
> Are you saying we shouldn't even check that we are writing valid notebooks to disk? That's the only thing we do check now

No, the on-disk format should always be validated.

> The in-memory format being discussed that should be different is never validated in the current system.

I don't think it't true, I might be wrong. But nbformat validation is in memory. 

I believe that the Browser <-> server channel to save/load notebook, should be able to use a private API that bypass the content manager, and which store more informations in the server memory[1] while the server is running. Things  like which callbacks/messages are attached to which cell; partial update.... etc. 

Then `save` action is "just" asking the server to export that to a `.ipynb` file, which do not require to send all on the wire. 
Trying to shoehorn all the client state on disk at each save is wrong[2]. 

I just think that these are actionable items, that can be private API, so no backward compatibility issues, and will fix a lot of problem that users currently have with the notebook.

[1] or on a scratch file, whatever I don't care. The server memory is just a  good candidate, because if your server is down, your kernels are probably down too, and so all these information are useless. 

[2] We saw that starting with widget reconnection.
